pandas/core/frame.py

```diff
@@ -1157,6 +1157,14 @@ def __len__(self) -> int:
         """
         return len(self.index)
 
+    @overload
+    def dot(self, other: Series) -> Series:
+        ...
+
+    @overload
+    def dot(self, other: Union[DataFrame, Index, ArrayLike]) -> Optional[DataFrame]:
```

Comment:
why is the return Optional?

Reply:
It's a (likely misguided) attempt to please mypy. There's no return statement after the raise at the end so technically returns None, although ofc we never get there

(I'm honestly not sure why what I have, without the optional return, isn't working - I think it _should_ be right)